File: templates/mobile/mobile/src/limits.ts
import limitsFixture from '../../limits.json';
import {
  LimitExceededError,
  checkCompactJsonUtf8Bytes,
  checkMeasurement,
  checkTrimmedUnicodeScalars,
} from '@baukit/data-contracts/limits';

const SUPPORTED_POLICY_VERSION = 1;

export interface LimitsPolicy {
  readonly $comment: string;
  readonly version: number;
  readonly text: { readonly max_characters: number };
  readonly collection: { readonly max_elements: number };
  readonly document: { readonly max_bytes: number };
  readonly rows: { readonly max_count: number };
  readonly body: { readonly max_bytes: number };
  readonly batch: { readonly max_items: number };
}

export type LimitReason =
  | 'text_too_long'
  | 'jsonb_too_large'
  | 'too_many_elements'
  | 'too_many_rows'
  | 'body_too_large'
  | 'batch_too_large';

export type JsonValue =
  null | boolean | number | string | readonly JsonValue[] | { readonly [key: string]: JsonValue };

export class LimitsPolicyError extends Error {}

export class LimitError extends Error {
  constructor(
    readonly reason: LimitReason,
    readonly field: string,
  ) {
    super(`Limit exceeded for ${field}: ${reason}`);
    this.name = 'LimitError';
  }
}

export function parseLimitsPolicy(value: unknown): LimitsPolicy {
  const policy = expectObject(value, 'limits');
  expectKeys(
    policy,
    ['$comment', 'version', 'text', 'collection', 'document', 'rows', 'body', 'batch'],
    'limits',
  );
  if (typeof policy['$comment'] !== 'string') {
    throw new LimitsPolicyError('limits.$comment must be a string');
  }
  if (policy['version'] !== SUPPORTED_POLICY_VERSION) {
    throw new LimitsPolicyError(`Unsupported limits policy version ${String(policy['version'])}`);
  }
  checkSection(policy, 'text', 'max_characters');
  checkSection(policy, 'collection', 'max_elements');
  checkSection(policy, 'document', 'max_bytes');
  checkSection(policy, 'rows', 'max_count');
  checkSection(policy, 'body', 'max_bytes');
  checkSection(policy, 'batch', 'max_items');
  return policy as unknown as LimitsPolicy;
}

export const LIMITS_POLICY = parseLimitsPolicy(limitsFixture);

export function checkText(field: string, value: string): void {
  mapLimitExceeded(field, 'text_too_long', () =>
    checkTrimmedUnicodeScalars(value, LIMITS_POLICY.text.max_characters),
  );
}

export function checkJsonDocument(field: string, value: JsonValue): void {
  mapLimitExceeded(field, 'jsonb_too_large', () =>
    checkCompactJsonUtf8Bytes(value, LIMITS_POLICY.document.max_bytes),
  );
}

export function checkCollection(field: string, count: number): void {
  checkCount(field, count, LIMITS_POLICY.collection.max_elements, 'too_many_elements');
}

export function checkRows(field: string, count: number): void {
  checkCount(field, count, LIMITS_POLICY.rows.max_count, 'too_many_rows');
}

export function checkBody(field: string, byteLength: number): void {
  checkCount(field, byteLength, LIMITS_POLICY.body.max_bytes, 'body_too_large');
}

export function checkBatch(field: string, count: number): void {
  checkCount(field, count, LIMITS_POLICY.batch.max_items, 'batch_too_large');
}

function expectObject(value: unknown, path: string): Record<string, unknown> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new LimitsPolicyError(`${path} must be an object`);
  }
  return value as Record<string, unknown>;
}

function expectKeys(
  value: Record<string, unknown>,
  expected: readonly string[],
  path: string,
): void {
  const actual = Object.keys(value).sort();
  const wanted = [...expected].sort();
  if (actual.length !== wanted.length || actual.some((key, index) => key !== wanted[index])) {
    throw new LimitsPolicyError(`${path} has unknown or missing fields`);
  }
}

function checkSection(
  policy: Record<string, unknown>,
  sectionName: string,
  valueName: string,
): void {
  const section = expectObject(policy[sectionName], `limits.${sectionName}`);
  expectKeys(section, [valueName], `limits.${sectionName}`);
  const value = section[valueName];
  if (!Number.isSafeInteger(value) || (value as number) <= 0) {
    throw new LimitsPolicyError(`limits.${sectionName}.${valueName} must be a positive integer`);
  }
}

function checkCount(field: string, actual: number, maximum: number, reason: LimitReason): void {
  if (!Number.isSafeInteger(actual) || actual < 0) {
    throw new RangeError(`${field} count must be a non-negative integer`);
  }
  mapLimitExceeded(field, reason, () => checkMeasurement(actual, maximum));
}

function mapLimitExceeded(field: string, reason: LimitReason, check: () => unknown): void {
  try {
    check();
  } catch (error) {
    if (error instanceof LimitExceededError) throw new LimitError(reason, field);
    throw error;
  }
}

File: templates/backend/backend/crates/__app__-domain/src/limits.rs
use std::sync::OnceLock;

use baukit_core::limits::{
    CompactJsonLimitError, LimitExceeded, check_compact_json_utf8_bytes, check_measurement,
    check_trimmed_unicode_scalars,
};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use thiserror::Error;

const SUPPORTED_POLICY_VERSION: u32 = 1;
const LIMITS_JSON: &str = include_str!("../../../../limits.json");

#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct LimitsPolicy {
    #[serde(rename = "$comment")]
    _comment: String,
    pub version: u32,
    pub text: TextLimits,
    pub collection: CollectionLimits,
    pub document: DocumentLimits,
    pub rows: RowLimits,
    pub body: BodyLimits,
    pub batch: BatchLimits,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct TextLimits {
    pub max_characters: usize,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct CollectionLimits {
    pub max_elements: usize,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct DocumentLimits {
    pub max_bytes: usize,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct RowLimits {
    pub max_count: usize,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct BodyLimits {
    pub max_bytes: usize,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct BatchLimits {
    pub max_items: usize,
}

#[derive(Debug, Error)]
pub enum LimitsPolicyError {
    #[error("limits policy is invalid: {0}")]
    InvalidJson(#[from] serde_json::Error),
    #[error("unsupported limits policy version {0}")]
    UnsupportedVersion(u32),
    #[error("limits policy value `{0}` must be greater than zero")]
    ZeroValue(&'static str),
}

pub fn parse_limits_policy(source: &str) -> Result<LimitsPolicy, LimitsPolicyError> {
    let policy: LimitsPolicy = serde_json::from_str(source)?;
    policy.validate()?;
    Ok(policy)
}

impl LimitsPolicy {
    fn validate(&self) -> Result<(), LimitsPolicyError> {
        if self.version != SUPPORTED_POLICY_VERSION {
            return Err(LimitsPolicyError::UnsupportedVersion(self.version));
        }
        for (name, value) in [
            ("text.max_characters", self.text.max_characters),
            ("collection.max_elements", self.collection.max_elements),
            ("document.max_bytes", self.document.max_bytes),
            ("rows.max_count", self.rows.max_count),
            ("body.max_bytes", self.body.max_bytes),
            ("batch.max_items", self.batch.max_items),
        ] {
            if value == 0 {
                return Err(LimitsPolicyError::ZeroValue(name));
            }
        }
        Ok(())
    }
}

pub fn limits_policy() -> &'static LimitsPolicy {
    static POLICY: OnceLock<LimitsPolicy> = OnceLock::new();
    POLICY.get_or_init(|| {
        parse_limits_policy(LIMITS_JSON)
            .unwrap_or_else(|error| panic!("embedded limits.json must be valid: {error}"))
    })
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum LimitReason {
    TextTooLong,
    JsonbTooLarge,
    TooManyElements,
    TooManyRows,
    BodyTooLarge,
    BatchTooLarge,
}

impl LimitReason {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::TextTooLong => "text_too_long",
            Self::JsonbTooLarge => "jsonb_too_large",
            Self::TooManyElements => "too_many_elements",
            Self::TooManyRows => "too_many_rows",
            Self::BodyTooLarge => "body_too_large",
            Self::BatchTooLarge => "batch_too_large",
        }
    }
}

#[derive(Clone, Copy, Debug, Eq, Error, PartialEq)]
#[error("limit exceeded for {field}: {reason:?}")]
pub struct LimitError {
    pub reason: LimitReason,
    pub field: &'static str,
}

impl LimitError {
    pub const fn code(&self) -> &'static str {
        self.reason.as_str()
    }
}

pub fn check_text(field: &'static str, value: &str) -> Result<(), LimitError> {
    map_limit_exceeded(
        field,
        LimitReason::TextTooLong,
        check_trimmed_unicode_scalars(value, limits_policy().text.max_characters),
    )
}

pub fn check_json_document(field: &'static str, value: &Value) -> Result<(), LimitError> {
    match check_compact_json_utf8_bytes(value, limits_policy().document.max_bytes) {
        Ok(_) => Ok(()),
        Err(CompactJsonLimitError::Limit(_)) => Err(LimitError {
            reason: LimitReason::JsonbTooLarge,
            field,
        }),
        Err(CompactJsonLimitError::Encoding(_)) => {
            unreachable!("serde_json::Value must serialize")
        }
    }
}

pub fn check_collection(field: &'static str, count: usize) -> Result<(), LimitError> {
    check_count(
        field,
        count,
        limits_policy().collection.max_elements,
        LimitReason::TooManyElements,
    )
}

pub fn check_rows(field: &'static str, count: usize) -> Result<(), LimitError> {
    check_count(
        field,
        count,
        limits_policy().rows.max_count,
        LimitReason::TooManyRows,
    )
}

pub fn check_body(field: &'static str, byte_length: usize) -> Result<(), LimitError> {
    check_count(
        field,
        byte_length,
        limits_policy().body.max_bytes,
        LimitReason::BodyTooLarge,
    )
}

pub fn check_batch(field: &'static str, count: usize) -> Result<(), LimitError> {
    check_count(
        field,
        count,
        limits_policy().batch.max_items,
        LimitReason::BatchTooLarge,
    )
}

fn check_count(
    field: &'static str,
    actual: usize,
    maximum: usize,
    reason: LimitReason,
) -> Result<(), LimitError> {
    map_limit_exceeded(field, reason, check_measurement(actual, maximum))
}

fn map_limit_exceeded(
    field: &'static str,
    reason: LimitReason,
    result: Result<baukit_core::limits::LimitMeasurement, LimitExceeded>,
) -> Result<(), LimitError> {
    result.map(|_| ()).map_err(|_| LimitError { reason, field })
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::*;

    #[test]
    fn loads_the_shared_product_fixture() {
        let parsed = parse_limits_policy(include_str!("../../../../limits.json"))
            .expect("shared limits fixture should parse");
        assert_eq!(limits_policy(), &parsed);
        assert_eq!(parsed.version, SUPPORTED_POLICY_VERSION);
    }

    #[test]
    fn rejects_an_unknown_version_and_zero_values() {
        let unknown_version = LIMITS_JSON.replace("\"version\": 1", "\"version\": 2");
        assert!(matches!(
            parse_limits_policy(&unknown_version),
            Err(LimitsPolicyError::UnsupportedVersion(2))
        ));

        let zero_text = LIMITS_JSON.replace("\"max_characters\": 200", "\"max_characters\": 0");
        assert!(matches!(
            parse_limits_policy(&zero_text),
            Err(LimitsPolicyError::ZeroValue("text.max_characters"))
        ));
    }

    #[test]
    fn rejects_unknown_policy_fields() {
        let unknown_field = LIMITS_JSON.replace("\"version\": 1", "\"version\": 1, \"extra\": 1");
        assert!(matches!(
            parse_limits_policy(&unknown_field),
            Err(LimitsPolicyError::InvalidJson(_))
        ));
    }

    #[test]
    fn checks_boundaries_and_stable_reason_codes() {
        let policy = limits_policy();
        assert_eq!(
            check_text("title", &"é".repeat(policy.text.max_characters)),
            Ok(())
        );
        assert_eq!(
            check_text("title", &"é".repeat(policy.text.max_characters + 1)),
            Err(LimitError {
                reason: LimitReason::TextTooLong,
                field: "title",
            })
        );
        assert_eq!(
            check_collection("entries", policy.collection.max_elements + 1)
                .expect_err("oversized collection should fail")
                .reason,
            LimitReason::TooManyElements
        );
        assert_eq!(
            check_rows("records", policy.rows.max_count + 1)
                .expect_err("too many rows should fail")
                .reason,
            LimitReason::TooManyRows
        );
        assert_eq!(
            check_body("request", policy.body.max_bytes + 1)
                .expect_err("oversized body should fail")
                .reason,
            LimitReason::BodyTooLarge
        );
        assert_eq!(
            check_batch("changes", policy.batch.max_items + 1)
                .expect_err("oversized batch should fail")
                .reason,
            LimitReason::BatchTooLarge
        );

        let reasons = [
            LimitReason::TextTooLong,
            LimitReason::JsonbTooLarge,
            LimitReason::TooManyElements,
            LimitReason::TooManyRows,
            LimitReason::BodyTooLarge,
            LimitReason::BatchTooLarge,
        ];
        assert_eq!(
            reasons.map(LimitReason::as_str),
            [
                "text_too_long",
                "jsonb_too_large",
                "too_many_elements",
                "too_many_rows",
                "body_too_large",
                "batch_too_large"
            ]
        );
        assert_eq!(
            serde_json::to_value(reasons).expect("reasons should serialize"),
            json!([
                "text_too_long",
                "jsonb_too_large",
                "too_many_elements",
                "too_many_rows",
                "body_too_large",
                "batch_too_large"
            ])
        );
    }

    #[test]
    fn measures_compact_json_in_utf8_bytes() {
        let value = json!({ "value": "é" });
        let byte_length = value.to_string().len();
        let oversized = json!({ "value": "x".repeat(limits_policy().document.max_bytes) });

        assert!(byte_length > value.to_string().chars().count());
        assert_eq!(
            check_json_document("metadata", &oversized)
                .expect_err("oversized JSON should fail")
                .reason,
            LimitReason::JsonbTooLarge
        );
    }
}
